### PR TITLE
[faster-rcnn] Non-Maximum Suppression

### PR DESCRIPTION
**Pull request by SeaOfOcean**
## What changes were proposed in this pull request?

add Non-Maximum Suppression

## How was this patch tested?

unit test

### CODE REVIEW

**Comment by SeaOfOcean** [comment]
pr validate pass

File: dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+
+  /**
+   *
+   * @param dets N*5
+   * @param thresh
+   * @return
+   */
+  def nms(dets: Tensor[Float], thresh: Float): Array[Int] = {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala]: as discussed. If the method is not used. please delete it.

File: dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  @transient var y2Buffer: Tensor[Float] = _
+
+  /**
+   *
+   * @param scores score tensor, size N
+   * @param bbox   bboxes 4*N
+   * @param thresh overlap thresh
+   * @return the length of indices after nms
+   */
+  def nms(scores: Tensor[Float], bbox: Tensor[Float], thresh: Float,
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala]: please check bbox size

File: dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    init(scores.nElement())
+    getAreas(x1, x2, y1, y2, areas)
+    // indices start from 1
+    val orderLength = getSortedScoreInds(scores, sortIndBuffer)
+    var indexLenth = 0
+    var i = 0
+    var curInd = 0
+
+    while (i < orderLength) {
+      def nmsOne(): Unit = {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala]: why def the method inside the loop?

File: dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+      nmsOne()
+      i += 1
+    }
+    indexLenth
+  }
+
+  private def getSortedScoreInds(scores: Tensor[Float], resultBuffer: Array[Int]): Int = {
+    // note that when the score is the same,
+    // the order of the indices are different in python and here
+    scores.topk(scores.nElement(), dim = 1, increase = false, result = sortedScores,
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala]: topk pass in a method into DimApply3. topk can be optimized special for contiguous case.

File: dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+   *
+   * @param scores score tensor, size N
+   * @param bbox   bboxes 4*N
+   * @param thresh overlap thresh
+   * @return the length of indices after nms
+   */
+  def nms(scores: Tensor[Float], bbox: Tensor[Float], thresh: Float,
+    indsBuffer: Array[Int]): Int = {
+    if (scores.nElement() == 0) return 0
+    val (x1, y1, x2, y2) = if (!bbox.isContiguous()) {
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala]: As discussed, we should rearrange the bbox to make it more cache friendly

**Comment by yiheng** [comment]
LGTM

**Pull request merged**